### PR TITLE
Update tests and documentation for Term::posts()

### PR DESCRIPTION
**Pull request by gchtr**
**Ticket**: #1917

## Issue

The documentation for `Term::posts()` could need some love. Currently, it’s quite hard what the function accepts without having a very close look at the code.

## Solution

Add more documentation.

## Impact

Everybody happy.

## Usage Changes

None.

## Considerations

For 2.x, we could still consider to change the 1st and 2nd parameters so that they don’t accept different values (like "either post type or class"). Again, I suggest an arguments array that we also applied in #1845.

## Testing

None.

### CODE REVIEW

File: lib/Term.php [hunk 1]
```diff
@@ -414,19 +414,48 @@ public function path() {
...
+	 *                                       arguments for the post query that this method is going.
+	 *                                       to perform. Default `10`.
+	 * @param string $post_type_or_class     Optional. Either the post type to get or the name of
+	 *                                       post class to use for the returned posts. Default
+	 *                                       `any`.
+	 * @param string $post_class             Optional. The name of the post class to use for the
+	 *                                       returned posts. Default `Timber\Post`.
 	 * @return array|bool|null
 	 */
 	public function posts( $numberposts_or_args = 10, $post_type_or_class = 'any', $post_class = '' ) {
```
Comment (palmiak) [inline lib/Term.php]: Much better variable names - but what would say about using them in `get_posts()` too?

**Comment by palmiak** [comment]
It's much clearer now 👍 

As I written in the comments I'm not sure about one thing in `get_posts` - if I would like to use:
```
term.posts( { posts_per_page: 5, orderby: 'menu_order' }, 'page', `myClass` )
```
and I have class `page` user might think it should works like this:
- post_type will be set to `page`
- postClass will be set to `myClass`
but it won't. It will use:
- post_type `post`
- postClass `page`

I think there should be some extra logic:
```
if ( $PostClass == '' && class_exists($post_type) ) {
	$PostClass = $post_type;
}
```
and
```
if ( $post_type != '' ){
    $args['post_type'] = $post_type;
} elseif ( $post_type == '' && !isset($args['post_type']) ) {
    $args['post_type'] = 'any';
} 
```
or am I overthinking this?

File: lib/Term.php (hunk 1 above)
Comment (gchtr) [reply lib/Term.php]: Good idea! I just updated variable names.

**Comment by gchtr** [comment]
@palmiak You are right! The logic that you described doesn’t work. I added some more tests to cover the different use cases. I think the reason that your use case fails is that in Timber, there’s a somewhat hidden logic that automatically uses a class for a post when it has the same name as the requested post type (as described in https://github.com/timber/timber/issues/799#issuecomment-192445207).

So when there’s a class `page` and we select the post type `page`, `Timber::get_posts()` will always use the `page` class, no matter what we pass in the third parameter of `Term::posts()`. Even if we would change the logic with your suggestions. But this is probably a use case most uses will very rarely run into. To me, that behavior seems intended, though I think it’s too much magic.

It’s also something that might change with object factories, according to @pascalknecht.

**Comment by palmiak** [comment]
@gchtr I fully understand. I think that for now we won't make it any better. Let's wait for @pascalknecht change and then we'll think should what to do next.

Great work as always Lukas 👍

**Comment by gchtr** [comment]
Some tests failed because of some autoloading issue. I submitted #1940 to fix that.

**Comment by gchtr** [comment]
@palmiak Now the tests work! I’d still need an approving review from you to be able to merge this 😉.

**Comment by palmiak** [comment]
I'll try to do this tomorrow or on monday

pt., 8 mar 2019, 21:19 użytkownik Lukas Gächter <notifications@github.com>
napisał:

**Comment by gchtr** [comment]
Could @jarednova or @palmiak approve this pull request so I can finally merge it? 😉

**Comment by jarednova** [comment]
Thanks @gchtr and @palmiak !

**Pull request merged**